kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+             " t:", duration, " ms")
+
+  -- network error or malformed DNS response
+  if not answers then
+    stats_increment(self.stats, key, "query_fail_nameserver")
+    err = "DNS server error: " .. tostring(err) .. ", took " .. duration .. " ms" 
+    table_insert(tries, { name .. ":" .. TYPE_TO_NAME[qtype], err })
+    return nil, err
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_increment(self.stats, key, answers.errstr and
+                                   "query_fail:" .. answers.errstr or
+                                   "query_succ")
+
+  -- DNS response error
+  if answers.errcode then
+    err = ("dns %s error: %s %s"):format(
+            answers.errcode < CACHE_ONLY_ERROR_CODE and "server" or "client",
+            answers.errcode, answers.errstr)
+    table_insert(tries, { name .. ":" .. TYPE_TO_NAME[qtype], err })
+  end
+
+  return answers
+end
+
+
+-- resolve all `name`s and return first usable answers
+local function resolve_query_names(self, names, qtype, tries)
+  local answers, err
+
+  for _, qname in ipairs(names) do
+    answers, err = resolve_query(self, qname, qtype, tries)
+
+    -- severe error occurred
+    if not answers then
+      return nil, err
+    end
+
+    if not answers.errcode then
+      return answers, nil, answers.ttl
+    end
+  end
+
+  -- not found in the search iteration
+  return answers, nil, answers.ttl
+end
+
+
+local function resolve_query_types(self, name, qtype, tries)
+  local names = search_names(name, self.resolv, self.hosts)
+  local answers, err, ttl
+
+  -- the specific type
+  if qtype ~= TYPE_A_AAAA then
+    return resolve_query_names(self, names, qtype, tries)
+  end
+
+  -- query A or AAAA
+  if self.enable_ipv4 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_A, tries)
+    if not answers or not answers.errcode then
+      return answers, err, ttl
+    end
+  end
+
+  if self.enable_ipv6 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_AAAA, tries)
+  end
+
+  return answers, err, ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype)
+  if premature then
+    return
+  end
+
+  local tries = setmetatable({}, TRIES_MT)
+  local answers = resolve_query_types(self, name, qtype, tries)
+  if answers and not answers.errcode then
+    log(DEBUG, PREFIX, "update stale DNS records: ", #answers)
+    self.cache:set(key, { ttl = answers.ttl }, answers)
+
+  else
+    log(DEBUG, PREFIX, "failed to update stale DNS records: ", tostring(tries))
+  end
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+  stats_increment(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype)
+  if not ok then
+    log(ALERT, PREFIX, "failed to start a timer to update stale DNS records: ", err)
+  end
+end
+
+
+local function check_and_get_ip_answers(name)
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
```

Comment:
Lua match may be slower than ngx.re.match, should we change it?

Reply:
or we could use kong api in kong/tools/ip.lua, let me check it

```
function _M.is_valid_ipv4(ipv4)
  return validate(ipv4, ipmatcher.parse_ipv4)
end
```

Reply:
will add a TODO comment